src/net_processing.cpp

```diff
@@ -1880,6 +1930,240 @@ void static ProcessOrphanTx(CConnman* connman, std::set<uint256>& orphan_work_se
+    }
+
+    bool index_in_sync = false;
+
+    uint256 prev_header;
+    if (start_height > 0) {
+        const CBlockIndex* prev_block = stop_index->GetAncestor(start_height - 1);
+        bool lookup_success = QueryFilterIndexWithRetry(
+            filter_index, index_in_sync,
+            [=, &prev_header]{ return filter_index->LookupFilterHeader(prev_block, prev_header); }
+        );
+        if (!lookup_success) {
+            return error("Failed to find block filter header in index: filter_type=%s, block_hash=%s",
+                         BlockFilterTypeName(filter_type), prev_block->GetBlockHash().ToString());
+        }
+    }
+
+    std::vector<uint256> filter_hashes;
+    bool lookup_success = QueryFilterIndexWithRetry(
+        filter_index, index_in_sync,
+        [=, &filter_hashes]{ return filter_index->LookupFilterHashRange(start_height, stop_index, filter_hashes); }
+    );
+    if (!lookup_success) {
+        return error("Failed to find block filter hashes in index: filter_type=%s, start_height=%d, stop_hash=%s",
+                     BlockFilterTypeName(filter_type), start_height, stop_hash.ToString());
+    }
+
+    CSerializedNetMsg msg = CNetMsgMaker(pfrom->GetSendVersion())
+        .Make(NetMsgType::CFHEADERS,
+              filter_type_ser,
+              stop_index->GetBlockHash(),
+              prev_header,
+              filter_hashes);
+    connman->PushMessage(pfrom, std::move(msg));
+
+    return true;
+}
+
+static bool ProcessGetCFCheckPt(CNode* pfrom, CDataStream& vRecv, const CChainParams& chain_params,
+                                CConnman* connman)
+{
+    uint8_t filter_type_ser;
+    uint256 stop_hash;
+
+    vRecv >> filter_type_ser >> stop_hash;
+
+    BlockFilterType filter_type = static_cast<BlockFilterType>(filter_type_ser);
+
+    const CBlockIndex* stop_index;
+    BlockFilterIndex* filter_index;
+    if (!PrepareBlockFilterRequest(pfrom, chain_params, filter_type, /*start_height=*/0, stop_hash,
+                                   /*max_height_diff=*/std::numeric_limits<uint32_t>::max(),
+                                   stop_index, filter_index)) {
+        // Return true because the issue with the invalid request has already been logged.
+        return true;
+    }
+
+    std::vector<uint256> headers(stop_index->nHeight / CFCHECKPT_INTERVAL);
+    {
+        std::lock_guard<std::mutex> _lock(active_chain_cf_headers_mtx);
+
+        bool index_in_sync = false;
+
+        // Populate headers.
+        int i = headers.size() - 1;
+        const CBlockIndex* block_index = stop_index;
+        for (; i >= 0; i--) {
+            uint32_t height = (i + 1) * CFCHECKPT_INTERVAL;
+            block_index = block_index->GetAncestor(height);
+
+            if (static_cast<size_t>(i) < active_chain_cf_headers.size() &&
+                active_chain_cf_headers[i].first == block_index) {
+                break;
+            }
+
+            // Filter header requested for stale block.
```

Comment:
What do you mean by this comment? Looking at the tests, it seems like we should be able to retrieve filters, headers and checkpoint headers on stale chains.